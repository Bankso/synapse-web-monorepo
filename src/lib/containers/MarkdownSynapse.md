Hard-coded markdown:
```jsx
    <MarkdownSynapse token={sessionToken} markdown='*markdown* given to the **component**' /> 
```
Or driven by a wiki page:
```jsx
  <MarkdownSynapse token={sessionToken} ownerId='syn12666371' wikiId='585317' />
```
Supports videos, images, and more.  No need to provide the wiki page ID for most entities:
```jsx
    <MarkdownSynapse token={sessionToken} ownerId='syn18142975'  />
```