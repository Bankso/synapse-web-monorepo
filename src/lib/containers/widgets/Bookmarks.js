import React from 'react'

class Bookmarks extends React.Component {

    constructor() {
        super()
        this.bookmarkRef = React.createRef()
        this.transformProps = this.transformProps.bind(this)
        this.state = {}
    }

    componentDidMount() {
        this.transformProps()
    }

    transformProps() {
        // find all links in the footnotes_html-- each of which contains a "text=[\d]"
        let copyFootnotes = String(this.props.footnotes)
        let linkOccurences = copyFootnotes.match(/text=\[.\]/g)
        if (!linkOccurences) {
            return
        }

        let linksFormatted = linkOccurences.map(
            (element, index ) => { 
                // grab only the [\d] pieces of the text
                return element.substring(element.indexOf("["), element.indexOf("]") + 1)
            }
        ).map(
            (element, index) => {
                // here bookmark is used so that the references can target this anchor tag 
                return `<span><div class="BookmarkWidget"><a id=${"bookmark" + index}>${element}</a></div></span>`
            }
        )

        // now all of the links are prepared to be inserted back into the original string
        // all link occurences have a single location in the original string-- we go through and then 
        // match of the link occurences to the spot it belongs into the html
        let i = 0
        let matches = copyFootnotes.replace(/(<span data-widgetparams=.*>)(&lt;Synapse widget&gt;)(<\/span>)/gm, 
            // replace using a function where p1,p2,p3 correspond to the matched groups from above
            // specifically removing the Synapse widget text and then putting instead of the anchor tag with the link
            // formatted text from above
            (match, p1, p2, p3, string) => {
                return [p1, linksFormatted[i++] , p3].join("")
            }
        )

        // create the dom element for this view and append to the ref
        let bookmarkFragment = document.createRange().createContextualFragment(matches)
        this.bookmarkRef.current.appendChild(bookmarkFragment)
    }

    render () {
        return (
            <React.Fragment>
                <hr/>
                <div ref={this.bookmarkRef}>
                </div>
            </React.Fragment>  
        )
    }
}

export default Bookmarks