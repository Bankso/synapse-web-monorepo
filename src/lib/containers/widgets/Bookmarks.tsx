import * as React from 'react'

type BookmarksProps = {
  footnotes: string
}

class Bookmarks extends React.Component<BookmarksProps, {}> {

  private bookmarkRef: React.RefObject<HTMLInputElement>

  constructor(props: BookmarksProps) {
    super(props)
    this.bookmarkRef = React.createRef()
    this.transformProps = this.transformProps.bind(this)
  }

  public componentDidMount(): void {
    this.transformProps()
  }

  public transformProps(): void {
        // find all links in the footnotes_html-- each of which contains a "text=[\d]"
    const copyFootnotes: string = String(this.props.footnotes)
    const linkOccurences: RegExpMatchArray | null = copyFootnotes.match(/text=\[.\]/g)
    if (!linkOccurences) {
      return
    }
    const linksFormatted = linkOccurences
            .map((element) => {
                // grab only the [\d] pieces of the text
                return element.substring(element.indexOf('['), element.indexOf(']') + 1)
            })
            .map((element, index) => {
                // here bookmark is used so that the references can target this anchor tag
              return `<span><span class="BookmarkWidget"><a id="bookmark${index}">${element}</a></span></span>`
            })
        // now all of the links are prepared to be inserted back into the original string
        // all link occurences have a single location in the original string-- we go through and then
        // match of the link occurences to the spot it belongs into the html
    let i: number = 0
    const matches = copyFootnotes.replace(
            /(<span data-widgetparams=.*>)&lt;Synapse widget&gt;(<\/span>)/g,
            // replace using a function where p1,p2 correspond to the matched groups from above
            // specifically removing the Synapse widget text and then putting instead of the anchor tag with the link
            // formatted text from above
            (match, p1, p2, original) => {
              const text = [p1, linksFormatted[i], p2].join('')
              i += 1
              return text
            }
        )
    // create the dom element for this view and append to the ref
    const bookmarkFragment = document.createRange().createContextualFragment(matches)
    // TS doesn't like possibly null values, bang operator ! will assert this is okay.
    this.bookmarkRef.current!.appendChild(bookmarkFragment)
  }

  public render() {
    return (
            <React.Fragment>
                <hr />
                <div ref={this.bookmarkRef} />
            </React.Fragment>
    )
  }
}

export default Bookmarks
