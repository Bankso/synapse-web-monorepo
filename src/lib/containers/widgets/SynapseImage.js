import React from 'react'
import PropTypes from 'prop-types'
import {getFiles, getEntity} from '../../utils/SynapseClient'

class SynapseImage extends React.Component {

    constructor() {
        super()
        this.getEntity = this.getEntity.bind(this)
        this.getSynapseFiles = this.getSynapseFiles.bind(this)
        this.compareById = this.compareById.bind(this)

        this.state = {
            isLoaded: false,
            preSignedURL: ""
        }

    }

    compareById(fileName, key) {
        return function(element) {
            return element[key] === fileName
        }
    }

    /**
     * Attach markdown to wiki attachments
     */
    matchToHandle(comparator, objectList) {
        if (objectList) {
            // make sure the files have loaded
            let filtered =  objectList.filter(comparator)
            return filtered
        }
    }

    getEntity() {
        const {token, synapseId } = this.props
        getEntity(token, synapseId).then(
            data => {
                let fileHandleAssociationList = [{
                    fileHandleId: data.dataFileHandleId,
                    associateObjectId: synapseId,
                    associateObjectType: "FileEntity"
                }]
                this.getSynapseFiles(fileHandleAssociationList, data.dataFileHandleId)
            }
        )
        
    }
    
    getSynapseFiles(fileHandleAssociationList, id) {
        // overload the method for two different use cases, one where
        // the image is attached to an entity and creates a list on the spot,
        // the other where list is passed in from componentDidMount in MarkdownSynapse
        let request = {
            requestedFiles: fileHandleAssociationList,
            includePreSignedURLs: true,
            includeFileHandles: false,
            includePreviewPreSignedURLs: false
        }
        getFiles(request, this.props.token)
        .then(data => {
            let match = this.matchToHandle(this.compareById(id, "fileHandleId"), data.requestedFiles);
            this.setState({
                preSignedURL: match[0].preSignedURL
            })  
        })
    }
    
    componentDidMount() {
        if (!this.props.token) {
            return
        }
        if (!this.props.hasOwnProperty("wikiId")) {
            this.getEntity()
        } else {
            const {fileName, fileResults} = this.props
            let match = this.matchToHandle(this.compareById(fileName, "fileName"), fileResults);
            let fileHandleAssociationList = [{
                fileHandleId: match[0].id,
                associateObjectId: this.props.wikiId,
                associateObjectType: "WikiAttachment"
            }]
            this.getSynapseFiles(fileHandleAssociationList, match[0].id)
        }
    }

    render () {
        let imgStyle = {}
        let {params} = this.props
        
        if (params && params.align) {
            imgStyle.float = params.align.toLowerCase()
        }
        
        if (this.state.isLoaded) {
            return (
                null
            )
        } else {
            return (
                <React.Fragment>
                    <img
                        alt="synapse"
                        className="img-fluid"
                        src={this.state.preSignedURL}
                        style={imgStyle}
                    />
                </React.Fragment>
            )
        }
    }

}

SynapseImage.propTypes = {
    wikiId: PropTypes.string,
    synapseId: PropTypes.string,
    token: PropTypes.string,
    fileName: PropTypes.string,
    fileResults: PropTypes.array
}

export default SynapseImage