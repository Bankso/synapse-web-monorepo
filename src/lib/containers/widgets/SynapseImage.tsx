import * as React from 'react'
import { BatchFileResult } from '../../utils/jsonResponses/BatchFileResult'
import { FileEntity } from '../../utils/jsonResponses/FileEntity'
import { FileHandle } from '../../utils/jsonResponses/FileHandle'
import { getEntity, getFiles } from '../../utils/SynapseClient'

type SynapseImageProps = {
  wikiId?: string
  synapseId?: string
  token?: string
  fileName?: string
  fileResults?: FileHandle []
  params: any
}

type SynapseImageState = {
  isLoaded: boolean
  preSignedURL: string
}

class SynapseImage extends React.Component<SynapseImageProps, SynapseImageState> {

  constructor(props: SynapseImageProps) {
    super(props)
    this.getEntity = this.getEntity.bind(this)
    this.getSynapseFiles = this.getSynapseFiles.bind(this)
    this.compareById = this.compareById.bind(this)
    this.state = {
      isLoaded: false,
      preSignedURL: ''
    }
  }

  public compareById(fileName: string, key: string) {
    return (element: any) => element[key] === fileName
  }

    /**
     * Attach markdown to wiki attachments
     */
  public matchToHandle(comparator: any , objectList: any) {
    if (objectList) {
            // make sure the files have loaded
      const filtered = objectList.filter(comparator)
      return filtered
    }
    return []
  }

  public getEntity() {
    const { token, synapseId } = this.props
    getEntity(token, synapseId!).then(
            // https://docs.synapse.org/rest/org/sagebionetworks/repo/model/FileEntity.html
            (data: FileEntity) => {
              const fileHandleAssociationList = [
                {
                  associateObjectId: synapseId,
                  associateObjectType: 'FileEntity',
                  fileHandleId: data.dataFileHandleId
                }
              ]
              this.getSynapseFiles(fileHandleAssociationList, data.dataFileHandleId)
            })
  }
  public getSynapseFiles(fileHandleAssociationList: any, id: string) {
        // overload the method for two different use cases, one where
        // the image is attached to an entity and creates a list on the spot,
        // the other where list is passed in from componentDidMount in MarkdownSynapse
    const request: any = {
      includeFileHandles: false,
      includePreSignedURLs: true,
      includePreviewPreSignedURLs: false,
      requestedFiles: fileHandleAssociationList
    }
    getFiles(request, this.props.token).then(
            (data: BatchFileResult) => {
              const match = this.matchToHandle(this.compareById(id, 'fileHandleId'), data.requestedFiles)
              this.setState({
                preSignedURL: match[0].preSignedURL
              })
            })
  }
  public componentDidMount() {
    if (!this.props.token) {
      return
    }
    if (!this.props.hasOwnProperty('wikiId')) {
      this.getEntity()
    } else {
      const { fileName, fileResults } = this.props
      const match = this.matchToHandle(this.compareById(fileName!, 'fileName'), fileResults!)
      const fileHandleAssociationList = [
        {
          associateObjectId: this.props.wikiId,
          associateObjectType: 'WikiAttachment',
          fileHandleId: match[0].id
        }
      ]
      this.getSynapseFiles(fileHandleAssociationList, match[0].id)
    }
  }
  public render() {
    const imgStyle: any = {}
    const { params } = this.props
    if (params && params.align) {
      imgStyle.float = params.align.toLowerCase()
    }
    if (this.state.isLoaded) {
      return null
    }
    return (
      <React.Fragment>
        <img
          alt="synapse"
          className="img-fluid"
          src={this.state.preSignedURL}
          style={imgStyle}
        />
      </React.Fragment>
    )
  }
}
export default SynapseImage
