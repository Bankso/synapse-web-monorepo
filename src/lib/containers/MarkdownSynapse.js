import React from "react";

import * as SynapseClient from '../utils/SynapseClient'
import Bookmarks from './widgets/Bookmarks'
import PropTypes from 'prop-types'
import SynapseImage from "./widgets/SynapseImage";
import katex from 'katex'

const uuidv4 = require('uuid/v4');

// Only because in the test enviornment there is an issue with importing
// react-plot which in turn imports mapboxgl which in turn defines a function
// that causes an error
let SynapsePlot
if (process.env.NODE_ENV !== "test") {
    import('./widgets/SynapsePlot').then(
        data => {
            SynapsePlot = data.default
        }
    )
}
let md = require('markdown-it')({html: true})
let markdownitSynapse = require('markdown-it-synapse')
let markdownSubAlt = require('markdown-it-sub-alt');
let markdownCenterText = require('markdown-it-center-text')
let markdownSynapseHeading = require('markdown-it-synapse-heading')
let markdownSynapseTable = require('markdown-it-synapse-table')
let markdownStrikethrough = require('markdown-it-strikethrough-alt')
let markdownContainer = require('markdown-it-container')
let markdownEmpahsisAlt = require('markdown-it-emphasis-alt')
let markdownInlineComments = require('markdown-it-inline-comments')
let markdownBr = require('markdown-it-br')
let sanitizeHtml = require('sanitize-html');
let synapseMath = require('markdown-it-synapse-math')

/**
 * Basic vanilla Markdownit functionality with latex support, synapse image support, plotly support
 *
 * @class Markdown
 * @extends {React.Component}
 */
class MarkdownSynapse extends React.Component {

    /**
     * Creates an instance of Markdown.
     * @param {*} props
     */
    constructor (props) {
        super(props)
        
        // markdownitSynapse wraps around md object and uses its own dependencies
        markdownitSynapse.init_markdown_it(
            md, markdownSubAlt, markdownEmpahsisAlt,
            markdownCenterText, markdownSynapseHeading, markdownSynapseTable,
            markdownStrikethrough, markdownContainer, markdownEmpahsisAlt,
            markdownInlineComments, markdownBr
        )

        const mathSuffix = ''
        // Update the internal md object with the wrapped synapse object
        md.use(markdownitSynapse, mathSuffix).use(synapseMath, mathSuffix)

        this.state = {
            md,
            text: '',
            fileHandles: null,
            newOwnerId: "",
            newWikiId: "",
            isLoggedIn: this.props.token !== "",
            errorMessage: ""
        }

        this.footnoteRef = React.createRef()
        this.markupRef = React.createRef()
        
        // handle widgets and math markdown
        this.processWidgets = this.processWidgets.bind(this)
        this.processWidgetOrDomElement = this.processWidgetOrDomElement.bind(this)
        this.processMath = this.processMath.bind(this)
        // handle init calls to get wiki related items
        this.getWikiAttachments = this.getWikiAttachments.bind(this)
        this.getWikiPageMarkdown = this.getWikiPageMarkdown.bind(this)
        
        // handle rendering widgets
        this.renderWidget = this.renderWidget.bind(this)
        this.renderSynapseButton = this.renderSynapseButton.bind(this)
        this.renderSynapseImage = this.renderSynapseImage.bind(this)
        this.renderSynapsePlot = this.renderSynapsePlot.bind(this)
        
        this.getErrorView = this.getErrorView.bind(this)
        this.createMarkup = this.createMarkup.bind(this)
        this.addBookmarks = this.addBookmarks.bind(this)
        this.handleMarkupClick = this.handleMarkupClick.bind(this)
    }

    /**
     * Given input text, generate markdown object to be passed onto inner html of some container.
     * @param {String} text The text being written in plain markdown
     * @returns {Object} Dictionary to be passed into dangerouslySetInnerHTML with markdown text
     */
    createMarkup(text) {
        let initText = this.state.md.render(text) 
        let cleanText = sanitizeHtml(initText, 
            {   
                allowedTags: [ 'span', 'code', 'h1', 'h2','h3', 'h4', 'h5', 'p', 'b', 'i', 'em', 'strong', 'a' ,'id',
            'table', 'tr', 'td', 'tbody', 'th', 'thead', "button", "div", "image", "ol", "ul", "li", "svg", "g",
            "br", "hr"],
                allowedAttributes: {
                    'a': [ 'href' ],
                    'span': ['*'],
                    'button': ['class'],
                    'div': ['class', 'style'],
                    "ul": ["class"],
                    "ol": ["class"],
                    "li": ["class"],
                    'table': ["class"],
                    'th': ['class'],
                    'thead': ['class']
                }
            }
        )
        return {__html: cleanText}
    }

    /**
     * Find all math identified elements of the form [id^=\"mathjax-\"]
     * (e.g. <dom element id="mathjax-10"> text </dom element>)
     * and transform them to their math markedown equivalents
     */
    processMath() {
        if (!this.markupRef.current) {
            return
        }
        // use regex to grab all elements
        let mathExpressions = this.markupRef.current.querySelectorAll("[id^=\"mathjax-\"]")
        // go through all obtained elements and transform them with katex
        mathExpressions.forEach(element => {
            katex.render(element.textContent, element, {throwOnError: false, delimiters: 
                [
                    {left: "$$", right: "$$", display: true},
                    {left: "\\(", right: "\\)", display: false},
                    {left: "\\[", right: "\\]", display: true}
                ]
            })
        });
    }

    /**
     * Process all the corresponding bookmark tags of the references made throughout the page
     *
     * @memberof MarkdownSynapse
     */
    addBookmarks() {

        markdownitSynapse.resetFootnotes()
        this.createMarkup(this.state.text)
        let footnotes_html = this.createMarkup(markdownitSynapse.footnotes()).__html

        if (footnotes_html.length > 0) {
            let bookmarks = <Bookmarks
                                footnotes={footnotes_html}>
                            </Bookmarks>
            return bookmarks
        }
    }
    
    /**
     * Call Synapse REST API to get AMP-AD wiki portal markdown as demo of API call
     */
    getWikiPageMarkdown() {
        if (!this.state.text) {
            SynapseClient.getEntityWiki(this.props.token, this.props.ownerId, this.props.wikiId)
                .then(data => {
                    // on success grab text and append to the default text
                    let initText = this.state.text;
                    this.setState({
                        text: initText + data.markdown
                    });
                    if (this.props.updateLoadState) {
                        this.props.updateLoadState({isLoading: false})
                    }
                }).catch(err => { 
                    console.log('Error on wiki markdown load\n', err);
                })
        }
        // else the wiki page was retrieved accordingly or it was passed down
        // as a prop
    }
              
    getWikiAttachments() {
        SynapseClient.getWikiAttachmentsFromEntity(this.props.token, this.props.ownerId, this.props.wikiId)
            .then(data => {
                this.setState({ fileHandles: data, errorMessage: "" });
            }).catch(err => { 
                this.setState({
                    errorMessage: err.reason
                })
                console.log("Error on wiki attachment load ", err)
            })
    }

    
    /**
     * If theres an error loading the wiki page show an informative message
     * likely a priveledge issue -- (e.g. not signed-in)
     * 
     * @returns view that presents error message on error, otherwise null
     */
    getErrorView() {
        if (this.state.errorMessage && this.props.errorMessageView) {
            return ( 
                <React.Fragment>
                    {React.cloneElement(this.props.errorMessageView, { message: this.state.errorMessage })}
                </React.Fragment>)
        }   
    }
    
    processWidgets() {
        // (<span data-widgetparams.*?span>) captures widgets
        let count = 1
        let markup = this.createMarkup(this.state.text).__html.replace(/<span id="wikiReference.*?<span data-widgetparams.*?span>/g, 
            () => {
                let current = count++
                return `<a href="" id="ref${current}">[${current}]</a>`
            }
        )
        let groups = markup.split(/(<span data-widgetparams.*?span>)/)
        if (groups.length > 0) {
            return this.processWidgetOrDomElement(groups)
        }
    }
    
    decodeXml(string) {
        let escaped_one_to_xml_special_map = {
            '&amp;': '&',
            '&quot;': '"',
            '&lt;': '<',
            '&gt;': '>'
        };
        
        return string.replace(/(&quot;|&lt;|&gt;|&amp;)/g,
        function(str, item) {
            return escaped_one_to_xml_special_map[item];
        });
    }
    
    processWidgetMappings(rawWidgetString, index) {
        let widgetstring = rawWidgetString.match(/data-widgetparams=("(.*?)")/)
        widgetstring = this.decodeXml(widgetstring[2])
        let questionIndex = widgetstring.indexOf("?")
        let widgetType = widgetstring.substring(0, questionIndex);
        let widgetparamsMapped = {};
        // map out params and their values
        widgetstring.substring(questionIndex + 1).split("&").forEach((keyPair) => {
            let key, value;
            [key, value] = keyPair.split("=");
            value = decodeURIComponent(value);
            widgetparamsMapped[key] = value;
        });
        return this.renderWidget(widgetType, widgetparamsMapped, index);
    } 
    
    processWidgetOrDomElement (widgetsToBe) {
        let widgets = []
        for(let i = 0; i < widgetsToBe.length;i++) {
            let text = widgetsToBe[i]
            if (text.indexOf("<span data-widgetparams") !== -1) {
                widgets.push(this.processWidgetMappings(text, i))
            } else {
                widgets.push(<span key={uuidv4()} dangerouslySetInnerHTML={{__html: text}}></span>)
            }
        }
        return widgets
    }
    
    renderWidget (widgetType, widgetparamsMapped) {
        switch (widgetType) {
            case "buttonlink":
                return this.renderSynapseButton(widgetparamsMapped)
            case "image":
                return this.renderSynapseImage(widgetparamsMapped);
            case "plot":
                return this.renderSynapsePlot(widgetparamsMapped);
            default:
                return
        }
    }
    
    renderSynapseButton(widgetparamsMapped) {
        return <a key={uuidv4()} href={widgetparamsMapped.url} className="btn btn-lg btn-info" role="button">{widgetparamsMapped.text}</a>
    }
    
    renderSynapsePlot(widgetparamsMapped) {
        return <SynapsePlot key={uuidv4()} token={this.props.token} ownerId={this.props.ownerId} wikiId={this.props.wikiId} widgetparamsMapped={widgetparamsMapped} />;
    }
    
    renderSynapseImage(widgetparamsMapped) {
        if (!this.state.fileHandles) {
            // ensure files are loaded
            return
        }
        
        if (widgetparamsMapped.fileName) {
            return <SynapseImage key={uuidv4()} token={this.props.token} fileName={widgetparamsMapped.fileName} wikiId={this.props.wikiId} fileResults={this.state.fileHandles.list} />;
        }
        else if (widgetparamsMapped.synapseId) {
            // elements with synapseIds have to have their resources loaded first, their not located
            // with the file attachnent list
            return <SynapseImage key={uuidv4()} token={this.props.token} synapseId={widgetparamsMapped.synapseId} />;
        }
    }
    
    componentDidMount() {
        if (this.props.markdown) {
            this.setState({
                text: this.props.markdown
            })
        }

        if (this.props.hasSynapseResources) {
            // get wiki attachments
            this.getWikiAttachments();
            this.getWikiPageMarkdown();
        }

        this.processMath()
    }

    // on component update find and re-render the math/widget items accordingly
    componentDidUpdate () {
        // we have to carefully update the component so it doesn't encounter an infinite loop
        if (this.props.token !== "" && !this.state.isLoggedIn) {
            // this is true when user just logged
            this.setState({isLoggedIn: true})
            // only if they didn't supply markdown should this happen
            if (this.props.hasSynapseResources) {
                this.getWikiAttachments()
                this.getWikiPageMarkdown()
            }
        }
        this.processMath()
    }

    handleMarkupClick(event) {
        event.preventDefault();
        if (event.target.tagName.toLowerCase() === 'a' && event.target.id.substring(0,3) === "ref") {  // check they clicked on anchor tag
            let referenceNumber = Number(event.target.id.substring(3)) // e.g. ref2 => '2'
            let goTo = this.footnoteRef.current.querySelector(`a#bookmark${referenceNumber - 1}`);
            try {
                goTo.scrollIntoView({
                    behavior: 'smooth',
                    block: 'center',
                    inline: 'center'
                });
            }
            catch (e) {
                console.log('error on scroll', e);
            }
        }
    }

    render() {
        return (
            <div>
                {this.getErrorView()}
                <span ref={this.markupRef} onClick={this.handleMarkupClick}>
                    {this.processWidgets()}
                </span>
                <div ref={this.footnoteRef}>
                    {this.addBookmarks()}
                </div>
            </div>
        )
    }
}

// Validate props passed to the component
MarkdownSynapse.propTypes = {
    // optional
    errorMessageView: PropTypes.element,

    // required props
    token : PropTypes.string.isRequired,

    ownerId: PropTypes.string,
    wikiId: PropTypes.string,

    markdown: PropTypes.string,
    hasSynapseResources: PropTypes.bool

}

MarkdownSynapse.defaultProps = {
    hasSynapseResources: true
}

export default MarkdownSynapse;