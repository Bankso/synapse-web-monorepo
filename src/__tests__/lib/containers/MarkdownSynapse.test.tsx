import * as React from 'react'
import { shallow } from 'enzyme'
import MarkdownSynapse, { MarkdownSynapseProps } from '../../../lib/containers/MarkdownSynapse'
import Bookmarks from '../../../lib/containers/widgets/Bookmarks'
import SynapseImage from '../../../lib/containers/widgets/SynapseImage'
import SynapsePlot from '../../../lib/containers/widgets/SynapsePlot'

const createShallowComponent = async (props: MarkdownSynapseProps, disableLifecycleMethods: boolean = false) => {
  const wrapper = await shallow(
    <MarkdownSynapse
      {...props}
    />,
    { disableLifecycleMethods }
  )
  const instance = wrapper.instance() as MarkdownSynapse
  return { wrapper, instance }
}

describe('it performs all functionality', () => {
  const SynapseClient = require('../../../lib/utils/SynapseClient')
  SynapseClient.getWikiAttachmentsFromEntity = jest.fn(() => Promise.resolve(['']))

  describe('renders with basic functionality ', () => {
    // test basic rendering capabilities
    //  - initial render
    //  - componentDidMount
    //  - componentDidUpdate
    const spyOnMath = jest.spyOn(MarkdownSynapse.prototype, 'processMath')
    beforeEach(() => {
      spyOnMath.mockReset()
    })

    it('renders without crashing', async () => {
      const props = { hasSynapseResources: false }
      const { wrapper } = await createShallowComponent(props, true)
      expect(wrapper).toBeDefined()
    })

    it('mounts correctly with markdown already loaded', async () => {
      const markdownPlaceholder = 'loremipsum....'
      const props = {
        markdown: markdownPlaceholder,
        hasSynapseResources: false
      }
      const { wrapper } = await createShallowComponent(props)

      expect(wrapper.state('text')).toEqual(markdownPlaceholder)
      expect(spyOnMath).toHaveBeenCalled()

    })

    it('mounts correctly with markdown NOT already loaded', async () => {
      const props = {
        wikiId: 'xxx', // placeholder
        ownerId: 'xxx' // placeholder
      }

      const markdownPlaceholder = 'text'

      // we only care to mock these functions and ensure they're called
      // Full functionality will get tested in the specific widget tests
      SynapseClient.getEntityWiki = jest.fn(() => { return Promise.resolve({ markdown: markdownPlaceholder }) })

      const spyOnWikiAttachmentsFromEntity = jest.spyOn(MarkdownSynapse.prototype, 'getWikiAttachments')
      const spyOnEntityWiki = jest.spyOn(MarkdownSynapse.prototype, 'getWikiPageMarkdown')

      // mount the component
      const { wrapper } = await createShallowComponent(props)

      // verify functions were called
      expect(wrapper.state('text')).toEqual(markdownPlaceholder)
      expect(spyOnWikiAttachmentsFromEntity).toHaveBeenCalledTimes(1)
      expect(spyOnEntityWiki).toHaveBeenCalledTimes(1)
    })

    it('runs componentDidUpdate correctly ', async () => {
      const props = {
        wikiId: 'xxx', // placeholder
        ownerId: 'xxx' // placeholder
      }

      // we only care to mock these functions and ensure they're called
      // Full functionality will get tested in the specific widget calls
      SynapseClient.getEntityWiki = jest.fn(() => Promise.resolve({ markdown: 'text' }))
      const { wrapper } = await createShallowComponent(props)

      // reset functions to test they were called inside of componentDidUpdate, from setProps
      SynapseClient.getEntityWiki = jest.fn(() => Promise.resolve({ markdown: 'text' }))
      SynapseClient.getWikiAttachmentsFromEntity = jest.fn(() => Promise.resolve(['']))

      await wrapper.setProps({
        token: '123'
      })

      // Note- verifying these API calls were made ensures that
      // getWikiAttachments and getWikiPageMarkdown were called
      expect(SynapseClient.getEntityWiki).toHaveBeenCalledTimes(1)
      expect(SynapseClient.getWikiAttachmentsFromEntity).toHaveBeenCalledTimes(1)
    })
  })

  describe('it renders an image widget', () => {
    it('renders an image from a synapseId ', async () => {
      SynapseClient.getEntityWiki = jest.fn(() =>
        Promise.resolve({
          markdown: '${image?synapseId=syn7809125&align=None&responsive=true}'
        })
      )
      const spyOnRenderImage = jest.spyOn(MarkdownSynapse.prototype, 'renderSynapseImage')
      const { wrapper } = await createShallowComponent({})
      expect(wrapper.find(SynapseImage)).toHaveLength(1)
      expect(spyOnRenderImage).toHaveBeenCalled()
    })

    it('renders an image from a file handleId ', async () => {
      SynapseClient.getEntityWiki = jest.fn(() =>
        Promise.resolve({
          markdown: '${image?fileName=joy%2Esvg&align=None&scale=100&responsive=true&altText=}'
        })
      )
      const spyOnRenderImage = jest.spyOn(MarkdownSynapse.prototype, 'renderSynapseImage')
      const { wrapper } = await createShallowComponent({})
      expect(wrapper.find(SynapseImage)).toHaveLength(1)
      expect(spyOnRenderImage).toHaveBeenCalled()
    })
  })

  describe('it renders a plot widget', () => {
    it('renders the SynapsePlot component', async () => {
      SynapseClient.getEntityWiki = jest.fn(() =>
        Promise.resolve({
          // tslint:disable-next-line:max-line-length
          markdown: '${plot?query=select "Age"%2C "Insol" from syn9872596&title=&type=BAR&barmode=GROUP&horizontal=false&showlegend=true}'
        })
      )
      const spyOnRenderPlot = jest.spyOn(MarkdownSynapse.prototype, 'renderSynapsePlot')
      const { wrapper } = await createShallowComponent({})
      expect(wrapper.find(SynapsePlot)).toHaveLength(1)
      expect(spyOnRenderPlot).toHaveBeenCalled()
    })
  })

  describe('it renders a synapse reference', () => {
    it('renders a synapse reference', async () => {
      // note- a reference is the anchor tag inside the text that links to the bookmark down below,
      // its an inline link
      SynapseClient.getEntityWiki = jest.fn(() => Promise.resolve({ markdown: '${reference?params}' }))
      const { wrapper } = await createShallowComponent({})
      // its unclear why its necessary to use .render() here, see https://github.com/airbnb/enzyme/issues/1233#issuecomment-406605838
      // the only alternative is to perform equality on .html() which is undesirable
      expect(wrapper.render().find('a#ref1')).toHaveLength(1)
    })
  })

  describe('it renders a bookmark', () => {
    it('renders a bookmark', async () => {
      // note - a bookmark is a corresponding citation for an inline reference, it provides a URL for
      // the reference.
      SynapseClient.getEntityWiki = jest.fn(() => Promise.resolve({ markdown: '${reference?text=google.com}' }))
      const { wrapper } = await createShallowComponent({})
      expect(wrapper.find(Bookmarks)).toHaveLength(1)
      expect(wrapper.render().find('a#bookmark0')).toHaveLength(1)
    })
  })

})
