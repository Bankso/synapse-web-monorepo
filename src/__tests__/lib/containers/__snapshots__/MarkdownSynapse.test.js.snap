// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <span
    onClick={[Function]}
  >
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </span>
  <div />
</div>
`;
